java-manta-client/src/main/java/com/joyent/manta/client/multipart/EncryptedMultipartManager.java

```diff
@@ -277,9 +285,15 @@ public ByteArrayOutputStream call(final long uploadedBytes) throws IOException {
                             }
                         }
                     });
+            final MantaMultipartUploadPart part = wrapped.uploadPart(upload.getWrapped(), partNumber, entity);
             encryptionState.setLastPartNumber(partNumber);
-            return wrapped.uploadPart(upload.getWrapped(), partNumber, entity);
+            return part;
         } finally {
+            if (encryptionState.getLastPartNumber() != partNumber && snapshot != null) {
+                // a snapshot was prepared but we didn't make it past uploadPart, rewind EncryptionState
+                RECORDER.rewind(encryptionState, snapshot);
```

Comment:
Can this possibly throw an exception?  I'm worried about anything happening in the same `finally` as the lock release.